source/agora/consensus/Reward.d

```diff
@@ -0,0 +1,413 @@
+import agora.common.Ensure;
+import agora.common.Types;
+import agora.utils.Log;
+
+import core.time;
+
+import std.algorithm;
+import std.array;
+import std.math;
+import std.range;
+
+/// struct to return the calculated block rewards for the `Validators` and the
+/// `Commons Budget`
+struct BlockRewards
+{
+    public Amount validator_rewards;
+    public Amount commons_budget_rewards;
+}
+
+/// number of seconds in a year of 365 days
+private const ulong yearOfSecs = 365 * 24 * 60 * 60;
+
+/// As defined in the Whitepaper: 27 coins every 5 seconds for first year (in Coin units)
+private const ulong firstYearValRewards = 27 * (yearOfSecs / 5) * 10_000_000L;
+
+/*******************************************************************************
+
+    Calculate Block Rewards for `Validators` and `Commons Budget`
+
+*******************************************************************************/
+
+public class Reward
+{
+    private Logger log;
+
+    /// As defined in the Whitepaper this is the total rewards to be sent to
+    /// Commons Budget over about 6.7 years
+    private const Amount totalCommonsBudgetRewards = 1_800_000_000.coins;
+
+    /// As defined in the Whitepaper this is the total rewards allocated to be
+    /// distributed to all the Validators over about 128 years
+    private const Amount totalValRewards = 2_700_000_000.coins;
+
+    /// Every 5 secs 50 coins are rewarded equals 10 per second (Block interval of 600 secs)
+    private const Amount commonsCoinsPerBlock = 6000.coins;
+
+    /// This is the Amount to pay the Commons Budget in a payout period before the allocation runs out
+    private Amount commonsPerPayout;
+
+    /// Number of blocks that rewards are paid for each payout (also the interval for payments)
+    private uint payout_period;
+
+    /// Duration of a block in seconds
+    private ulong block_interval_sec;
+
+    /// Number of Coinbase transactions per year
+    private ulong payouts_per_year;
+
+    /***************************************************************************
+
+        Construct `Reward` object which encapsulates block reward calculations.
+
+        Params:
+            payout_period = how many blocks are included for a reward payout
+            block_interval_sec = time span for each block
+
+    ***************************************************************************/
+
+    public this (in uint payout_period, in Duration block_interval)
+    {
+        this.log = Logger(__MODULE__);
+
+        this.payout_period = payout_period;
+        this.block_interval_sec = block_interval.total!"seconds";
+
+        // Calculate payouts per year
+        this.payouts_per_year = yearOfSecs / (this.payout_period * this.block_interval_sec);
+
+        // Calculate amount that would be paid to Commons Budget for full payout period
+        // if the allocation is still sufficient
+        this.commonsPerPayout = commonsCoinsPerBlock; commonsPerPayout.mul(this.payout_period);
```

Comment:
Not checking the return value defeats the point of not having operator overload in the first place!

I suppose you need something like this ?
```D
if (commonsPerPayout.mul(this.payout_period))
    throw new Exception("The configured payout period is too long");
```

Reply:
Updated